keps/1833-metrics-for-local-queue/README.md

```diff
@@ -0,0 +1,167 @@
+# KEP-1833: Enable Prometheus Metrics for Local Queues
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+- [Design Details](#design-details)
+  - [API changes:](#api-changes)
+  - [List of metrics for Local Queues:](#list-of-metrics-for-local-queues)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+<!-- /toc -->
+
+## Summary
+
+The enhancement aims to introduce the exposure of local queue metrics to users, providing detailed insights into workload 
+processing specific to individual namespaces / tenants.
+
+## Motivation
+
+Metrics related to local queues are invaluable for batch users (usually with namespace-scoped permissions), as they provide 
+essential visibility and historical trends about their workloads. Currently, metrics are available for ClusterQueues but 
+not for namespace-scoped batch users. Cluster queue metrics are often ineffective for batch users and namespace admins 
+because they are global and cannot be filtered by namespaces. Furthermore, accessing cluster-scoped metrics 
+in secured Prometheus instances is generally restricted to cluster admin users with cluster level permissions across all namespaces and 
+tenants. This restriction makes it challenging for batch users to obtain the specific metrics they need for effective workload 
+management and to gain insights into their workloads within their limited scope of access.
```

Comment:
Could this KEP also specify how we will prevent users with insufficient permission accessing metrics they shouldn't be able to?

Reply:
If this proposal is not suggesting how to prevent access from namespaces that shouldn't have the permission, then I would remove this phrase.

Reply:
I agree, maybe we could narrow the scope of this KEP @varshaprasad96 ? At first glance managing permissions seems to be challenging, or it would require external mechanism

Reply:
Sorry for the late reply! I've been considering various options, and it's clear that publishing local queue metrics in each namespace could lead to complications. This approach would require multiple endpoints with respective Service Monitors or, if using a single central Service Monitor, the correct RBAC setup to allow client access. The centralised approach could still poses some issues, such as namespace admins potentially being able to view metrics from other namespaces.

One potential solution for cluster admins is to scaffold out a Service Monitor (SM) with metrics [labeling](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#metric_relabel_configs) for specific namespaces from the same service endpoint. This would enable a common service endpoint for all local queues. Admins could then provide a service account with right RBAC to specific batch user, restricting their access to that particular service monitor. 

However, this solution is difficult to implement in Kueue right away by figuring out the right set of scaffolds. This could probably be a customisation which can be documented for now. 

That being said, given the complexity, this topic may probably warrant a further brainstorming and deserves a separate KEP. For now, I'll remove this from the current discussion and update the document to indicate that for this iteration metrics will be exported in the controller namespace, alongside cluster queue metrics, at the same endpoint. Does that sound reasonable, or would we want to a separate endpoint for all local queue metrics?

Reply:
Yes it does.

You can remove the per-namespace topic from the motivation and add a note in non-goals.